src/go/util/httppattern/uri_template.go

```diff
@@ -0,0 +1,411 @@
+			if p.inVariable {
+				return markVariableHasWildcardPath()
+			}
+			return true
+		} else {
+			p.segments = append(p.segments, "*")
+			return true
+		}
+	case '{':
+		return p.parseVariable()
+	default:
+		return p.parseLiteralSegment()
+	}
+}
+
+// Variable = "{" FieldPath [ "=" Segments ] "}" ;
+func (p *parser) parseVariable() bool {
+	if !p.consume('{') {
+		return false
+	}
+	if !p.startVariable() {
+		return false
+	}
+	if !p.parseFieldPath() {
+		return false
+	}
+	if p.consume('=') {
+		if !p.parseSegments() {
+			return false
+		}
+	} else {
+		p.segments = append(p.segments, "*")
+	}
+
+	if !p.endVariable() {
+		return false
+	}
+	if !p.consume('}') {
+		return false
+	}
+
+	return true
+}
+
+// FieldPath = IDENT { "." IDENT } ;
+func (p *parser) parseFieldPath() bool {
+	if !p.parseIdentifier() {
+		return false
+	}
+
+	for p.consume('.') {
+		if !p.parseIdentifier() {
+			return false
+		}
+	}
+	return true
+}
+
+// Verb     = ":" LITERAL ;
+func (p *parser) parseVerb() bool {
+	if !p.consume(':') {
+		return false
+	}
+	verb, result := p.parseLiteral()
+	if !result {
+		return false
+	}
+
+	p.verb = verb
+	return true
+}
+
+func (p *parser) parseIdentifier() bool {
+	addFieldIdentifier := func(id string) bool {
+		if p.inVariable && len(p.variables) > 0 {
+			p.currentVariable().FieldPath = append(p.currentVariable().FieldPath, id)
+			return true
+		} else {
+			// something's wrong we're not in a variable
+			return false
+		}
+	}
+
+	var idf bytes.Buffer
+	// Initialize to false to handle empty literal.
+	result := false
+	for p.nextChar() {
+		switch c := p.currentChar(); c {
+		case '.':
+			fallthrough
+		case '}':
+			fallthrough
+		case '=':
+			return result && addFieldIdentifier(idf.String())
+		default:
+			p.consume(c)
+			idf.WriteByte(c)
+			break
+		}
+		result = true
+	}
+
+	return result && addFieldIdentifier(idf.String())
+}
+
+func (p *parser) parseLiteral() (string, bool) {
+	var buffer bytes.Buffer
+	if !p.ensureCurrent() {
+		return "", false
+	}
+
+	// Initialize to false in case we encounter an empty literal.
+	result := false
+
+	for {
+		switch c := p.currentChar(); c {
+		case '/':
+			fallthrough
+		case ':':
+			fallthrough
+		case '}':
+			return buffer.String(), result
+		default:
+			p.consume(c)
+			buffer.WriteByte(c)
+			break
+		}
+		result = true
+
+		if !p.nextChar() {
+			break
+		}
+	}
+
+	return buffer.String(), result
+}
+
+func (p *parser) parseLiteralSegment() bool {
+	l, result := p.parseLiteral()
+	if !result {
+		return false
+	}
+
+	p.segments = append(p.segments, l)
+	return true
+}
+
+func (p *parser) consume(c byte) bool {
+	if p.tb >= p.te && !p.nextChar() {
+		return false
+	}
+
+	if p.currentChar() != c {
+		return false
+	}
+
+	p.tb += 1
+	return true
+}
+
+func (p *parser) consumeAllInput() bool {
+	return p.tb >= len(p.input)
+}
+
+func (p *parser) currentChar() byte {
+	if p.tb < p.te && p.te <= len(p.input) {
+		return p.input[p.te-1]
+	}
+
+	return InvalidChar
```

Comment:
Callers of `currentChar()` do not handle this `InvalidChar`, is that expected? Will parsing succeed with these 0 characters in the template now?

Reply:
Yes.

Nop, I think ""  are in the testcases.